### ISSUE TITLE
Issues with calling .shift() on a rest argument.

### ISSUE THREAD

**Issue by zenlambda** [babel/babel#918]
I think the following test should pass, at the moment it does not. 

```
var TEST_ARRAY = [1,2];
 
function testShift(...rest) {
  var first = rest.shift(),
    second = rest[0];
 
  return second;
}
 
assert.equal(TEST_ARRAY[1], testShift(1,2));
```

This seems to be because babel is replacing references to 'rest' subscripts with references to arguments[n+num_non_vargs], which seems strange. According to [1] 'The rest parameter is a *real Array* and it is a snapshot of the arguments passed to the function. ' (emphasis mine), so all the regular methods should work.

Here is babel's current output:

```
   1 | "use strict";
   2 | 
   3 | var TEST_ARRAY = [1, 2];
   4 | 
   5 | function testShift() {
   6 |   for (var _len = arguments.length, rest = Array(_len), _key = 0; _key < _len; _key++) {
   7 |     rest[_key] = arguments[_key];
   8 |   }
   9 | 
  10 |   var first = rest.shift(),
  11 |       second = arguments[0];
  12 | 
  13 |   return second;
  14 | }
  15 | 
  16 | assert.equal(TEST_ARRAY[1], testShift(1, 2));
```

What do you think?

[1] http://tc39wiki.calculist.org/es6/rest-parameters/

**Comment by neVERberleRfellerER** [comment]
@sebmck This is consequence of inlining in https://github.com/babel/babel/commit/751557aef19c34969be3d2ab788b9e63c175ea5b . rest is replaced in `rest[0]` but not in `rest.shift()`. I'll look at this.

**Comment by sebmck** [comment]
@neVERberleRfellerER I'm not going to be available for the rest of the day so I'll quickly fix it and push out a patch.

**Comment by neVERberleRfellerER** [comment]
@sebmck Ok. My fix would take little longer (probably, you are very good at fixing things). I do not see other way than building list of candidates as it used to be before inlining.

**Comment by neVERberleRfellerER** [comment]
@sebmck Your fix does not work in following case


    function f(x, y, ...args) {
        console.log(args[0]);
        args.pop();
        console.log(args[1]);
    }

**Comment by sebmck** [comment]
Welp, need to run. Mobile internet in Australia is shit.

On Sunday, 1 March 2015, neVERberleRfellerER <notifications@github.com>
wrote:

> @sebmck <https://github.com/sebmck> Your fix does not work in following
> case
>
> function f(x, y, ...args) {
>     console.log(args[0]);
>     args.pop();
>     console.log(args[1]);
> }
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/babel/babel/issues/918#issuecomment-76556358>.
>


-- 
Sebastian McKenzie

**Comment by neVERberleRfellerER** [comment]
@sebmck Ok. Let's hope this bug will not bite somebody...

**Comment by sebmck** [comment]
Probably will. I'll try and push out a fix when I'm on the road,
hopefully...

On Sunday, 1 March 2015, neVERberleRfellerER <notifications@github.com>
wrote:

> @sebmck <https://github.com/sebmck> Ok. Let's hope this bug will not bite
> somebody...
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/babel/babel/issues/918#issuecomment-76556535>.
>


-- 
Sebastian McKenzie

**Comment by sebmck** [comment]
Fixed as of 4.6.3, thanks!
